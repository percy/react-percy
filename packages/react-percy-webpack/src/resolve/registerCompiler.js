export default function registerCompiler(compiler) {
  if (!compiler) {
    return;
  }

  if (typeof compiler === 'string') {
    // eslint-disable-next-line global-require, import/no-dynamic-require
    require(compiler);
  } else if (!Array.isArray(compiler)) {
    // eslint-disable-next-line global-require, import/no-dynamic-require
    compiler.register(require(compiler.module));
  } else {
    for (let i = 0; i < compiler.length; i++) {
      try {
        registerCompiler(compiler[i]);
        break;
      } catch (e) {
        // continue trying subsequent compilers
      }
    }
  }
}
