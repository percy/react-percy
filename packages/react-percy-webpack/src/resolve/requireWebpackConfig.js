import * as path from 'path';
import chalk from 'chalk';
import fs from 'fs';
import interpret from 'interpret';
import getExtension from './getExtension';
import registerCompiler from './registerCompiler';

export default function requireWebpackConfig(webpackConfigPath) {
  const resolvedWebpackConfigPath = path.resolve(webpackConfigPath);
  if (!fs.existsSync(resolvedWebpackConfigPath)) {
    throw new Error(
      chalk.red('Webpack config file ') +
        chalk.red.bold(webpackConfigPath) +
        chalk.red(' could not be found.'),
    );
  }

  const extension = getExtension(resolvedWebpackConfigPath);
  registerCompiler(interpret.extensions[extension]);

  // eslint-disable-next-line global-require, import/no-dynamic-require
  return require(resolvedWebpackConfigPath);
}
