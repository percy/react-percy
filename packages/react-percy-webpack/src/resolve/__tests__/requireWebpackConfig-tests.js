import * as path from 'path';
import fs from 'fs';
import interpret from 'interpret';
import registerCompiler from '../registerCompiler';
import requireWebpackConfig from '../requireWebpackConfig';

jest.mock('fs', () => ({
  existsSync: jest.fn(),
}));

jest.mock('interpret', () => ({
  extensions: {
    '.foo.js': {
      'interpret-foo': 'mock',
    },
  },
}));

jest.mock('../getExtension', () => () => '.foo.js');
jest.mock('../registerCompiler', () => jest.fn());

const configPath = 'webpack.config.foo.js';

beforeEach(() => {
  jest.resetModules();

  fs.existsSync.mockReturnValue(true);
});

const givenWebpackConfig = (mockConfig = {}) => {
  jest.mock(path.resolve(configPath), () => mockConfig, { virtual: true });
};

it('throws when the webpack config cannot be found on disk', () => {
  fs.existsSync.mockReturnValue(false);

  expect(() => requireWebpackConfig(configPath)).toThrow();
});

it('registers the necessary compilers before loading the config', () => {
  givenWebpackConfig();

  requireWebpackConfig(configPath);

  expect(registerCompiler).toHaveBeenCalledWith(interpret.extensions['.foo.js']);
});

it('returns webpack config', () => {
  givenWebpackConfig({
    entry: {
      foo: 'bar',
    },
    module: {
      loaders: [],
    },
  });

  const config = requireWebpackConfig(configPath);

  expect(config).toEqual({
    entry: {
      foo: 'bar',
    },
    module: {
      loaders: [],
    },
  });
});
