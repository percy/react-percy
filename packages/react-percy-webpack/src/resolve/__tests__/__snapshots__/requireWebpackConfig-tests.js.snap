// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws when the webpack config cannot be found on disk 1`] = `"[31mWebpack config file [39m[31m[1mwebpack.config.foo.js[22m[39m[31m could not be found.[39m"`;
